File: src/Moka.Red.Primitives/Media/MokaMediaGallery.razor.cs
using System.Globalization;
using Microsoft.AspNetCore.Components;
using Microsoft.AspNetCore.Components.Web;
using Moka.Red.Core.Base;
using Moka.Red.Core.Enums;
using Moka.Red.Core.Utilities;

namespace Moka.Red.Primitives.Media;

/// <summary>
///     Grid of media thumbnails with hover overlay. Emits click events for lightbox integration.
///     Clickable thumbnails are buttons, and the lightbox is a modal dialog that takes focus and
///     locks the page's scrolling while it is open.
/// </summary>
public partial class MokaMediaGallery : MokaVisualComponentBase
{
	private const string GalleryModule = "./_content/Moka.Red.Primitives/Media/MokaMediaGallery.razor.js";

	private int _lightboxIndex;
	private MokaMediaItem? _lightboxItem;
	private bool _lightboxShown;
	private ElementReference _root;
	private ElementReference _lightbox;

	/// <summary>Collection of media items to display. Required.</summary>
	[Parameter]
	[EditorRequired]
	public IReadOnlyList<MokaMediaItem> Items { get; set; } = [];

	/// <summary>Number of grid columns. Default 3.</summary>
	[Parameter]
	public int Columns { get; set; } = 3;

	/// <summary>Gap between thumbnails. Default Sm.</summary>
	[Parameter]
	public MokaSpacingScale Gap { get; set; } = MokaSpacingScale.Sm;

	/// <summary>Whether to show a hover overlay with zoom icon. Default true.</summary>
	[Parameter]
	public bool ShowOverlay { get; set; } = true;

	/// <summary>Callback when a media item is clicked.</summary>
	[Parameter]
	public EventCallback<MokaMediaItem> OnItemClick { get; set; }

	/// <summary>
	///     Whether clicking an image opens a built-in fullscreen lightbox. Default true.
	///     When false, only fires OnItemClick, and you handle the preview yourself.
	/// </summary>
	[Parameter]
	public bool Lightbox { get; set; } = true;

	/// <inheritdoc />
	protected override string RootClass => "moka-media-gallery";

	/// <inheritdoc />
	protected override string CssClass => new CssBuilder(RootClass)
		.AddClass("moka-fill-width")
		.AddClass(Class)
		.Build();

	// The root only lays out the grid and draws nothing, so a radius there would not show. Every
	// thumbnail draws its own box and takes it instead. Margin and padding stay on the root.

	/// <inheritdoc />
	protected override string? CssStyle => new StyleBuilder()
		.AddStyle("margin", ResolvedMargin)
		.AddStyle("padding", ResolvedPadding)
		.AddStyle("grid-template-columns", $"repeat({Columns}, 1fr)")
		.AddStyle("gap", MokaEnumHelpers.ToCssValue(Gap))
		.AddStyle(Style)
		.Build();

	private string? ItemStyle => new StyleBuilder()
		.AddStyle("border-radius", ResolvedRounding)
		.Build();

	// A thumbnail that does nothing on click is not a button.
	private bool IsInteractive => Lightbox || OnItemClick.HasDelegate;

	private static string GetThumbnailSrc(MokaMediaItem item) =>
		item.ThumbnailSrc ?? item.Src;

	// A thumbnail button is named by its image's alt text. Without one it falls back to the
	// caption, then to its place in the gallery.
	private string? ItemLabel(MokaMediaItem item, int index)
	{
		if (!string.IsNullOrWhiteSpace(item.Alt))
		{
			return null;
		}

		return string.IsNullOrWhiteSpace(item.Caption)
			? string.Create(CultureInfo.CurrentCulture, $"Image {index + 1} of {Items.Count}")
			: item.Caption;
	}

	/// <inheritdoc />
	protected override bool ShouldRender() => true;

	/// <inheritdoc />
	protected override async Task OnAfterRenderAsync(bool firstRender)
	{
		// Focus moves into the lightbox when it opens and back to a thumbnail when it closes: the
		// one on show, which is where the arrows left the user. The script also locks the page's
		// scrolling on open and gives the lock back on close.
		bool shown = _lightboxItem is not null;
		if (shown == _lightboxShown)
		{
			return;
		}

		_lightboxShown = shown;
		if (shown)
		{
			await SafeModuleInvokeVoidAsync(GalleryModule, "openLightbox", _root, _lightbox);
		}
		else
		{
			await SafeModuleInvokeVoidAsync(GalleryModule, "closeLightbox", _root, _lightboxIndex);
		}
	}

	private async Task HandleItemClick(MokaMediaItem item, int index)
	{
		if (Lightbox)
		{
			_lightboxItem = item;
			_lightboxIndex = index;
		}

		if (OnItemClick.HasDelegate)
		{
			await OnItemClick.InvokeAsync(item);
		}
	}

	private void CloseLightbox() => _lightboxItem = null;

	private void LightboxPrev()
	{
		if (_lightboxIndex > 0)
		{
			_lightboxIndex--;
			_lightboxItem = Items[_lightboxIndex];
		}
	}

	private void LightboxNext()
	{
		if (_lightboxIndex < Items.Count - 1)
		{
			_lightboxIndex++;
			_lightboxItem = Items[_lightboxIndex];
		}
	}

	/// <inheritdoc />
	protected override async ValueTask DisposeAsyncCore()
	{
		// A gallery removed with its lightbox open never renders it closed, and the page would stay
		// locked. Nothing is sent when the script never heard of an open lightbox.
		if (_lightboxShown)
		{
			_lightboxShown = false;
			await SafeModuleInvokeVoidAsync(GalleryModule, "disposeLightbox", _root);
		}

		await base.DisposeAsyncCore();
	}

	private void HandleLightboxKeyDown(KeyboardEventArgs e)
	{
		if (e.Key == "Escape")
		{
			CloseLightbox();
			return;
		}

		// Alt+Left and Alt+Right are the browser's back and forward, not a change of image.
		if (e.AltKey || e.CtrlKey || e.MetaKey)
		{
			return;
		}

		switch (e.Key)
		{
			case "ArrowLeft":
				LightboxPrev();
				break;
			case "ArrowRight":
				LightboxNext();
				break;
		}
	}
}

File: src/Moka.Red.Primitives/Media/MokaMediaGallery.razor.js
/**
 * MokaMediaGallery JS module - focus handling and the scroll lock for the lightbox, which is a
 * modal dialog.
 *
 * Opening moves focus into the lightbox, keeps Tab inside it and locks the page's scrolling.
 * Closing gives the lock back and hands focus to the thumbnail of the image on show. Escape and
 * the arrow keys themselves are handled in .NET.
 *
 * The lock is Core's, the one dialogs and bottom sheets take through moka-dialog.js. The relative
 * import resolves against this module's own URL to the same module, so they share one count and
 * a lightbox closing over a dialog leaves the dialog's lock in place.
 */

import { lockBodyScroll, unlockBodyScroll } from '../../Moka.Red.Core/moka-scroll-lock.js';

const FOCUSABLE = 'button:not([disabled]), [href], input:not([disabled]), [tabindex]:not([tabindex="-1"])';

// Keys that would scroll the page behind the lightbox. The .NET handler still gets them.
const SCROLL_KEYS = ['ArrowLeft', 'ArrowRight', 'ArrowUp', 'ArrowDown', 'PageUp', 'PageDown', 'Home', 'End'];

// Gallery root -> MutationObserver of its open lightbox. A gallery in here also holds one scroll
// lock, given back when it leaves.
const openLightboxes = new WeakMap();

function trapTab(lightbox, e) {
	const focusable = Array.from(lightbox.querySelectorAll(FOCUSABLE));
	if (focusable.length === 0) {
		e.preventDefault();
		return;
	}

	const first = focusable[0];
	const last = focusable[focusable.length - 1];
	const active = document.activeElement;

	if (e.shiftKey && (active === first || active === lightbox)) {
		e.preventDefault();
		last.focus();
	} else if (!e.shiftKey && active === last) {
		e.preventDefault();
		first.focus();
	}
}

// Stops watching the gallery's lightbox and gives back its scroll lock, once per open.
function release(root) {
	const observer = openLightboxes.get(root);
	if (!observer) return;

	observer.disconnect();
	openLightboxes.delete(root);
	unlockBodyScroll();
}

/**
 * Takes focus into a lightbox that has just opened, keeps it there, and locks the page's scrolling.
 * @param {HTMLElement} root - The gallery's root element.
 * @param {HTMLElement} lightbox - The lightbox element, role="dialog".
 */
export function openLightbox(root, lightbox) {
	if (!root || !lightbox) return;

	// Opened again without a close in between: the gallery already holds a lock, and a second one
	// would never be given back.
	const previous = openLightboxes.get(root);
	if (previous) {
		previous.disconnect();
	} else {
		lockBodyScroll();
	}

	lightbox.addEventListener('keydown', e => {
		if (e.key === 'Tab') {
			trapTab(lightbox, e);
			return;
		}

		// With Alt, Ctrl or Cmd these are browser shortcuts (Alt+Left is back), so they stay.
		if (e.altKey || e.ctrlKey || e.metaKey) return;

		if (SCROLL_KEYS.includes(e.key) || (e.key === ' ' && e.target === lightbox)) {
			e.preventDefault();
		}
	});

	// The previous and next buttons disappear at either end. When the one that had focus goes,
	// focus returns to the lightbox, where Escape and the arrows still reach the .NET handler.
	const observer = new MutationObserver(() => {
		if (lightbox.isConnected && !lightbox.contains(document.activeElement)) {
			lightbox.focus({ preventScroll: true });
		}
	});
	observer.observe(lightbox, { childList: true, subtree: true });
	openLightboxes.set(root, observer);

	lightbox.focus({ preventScroll: true });
}

/**
 * Called after the lightbox has left the page. Scrolling comes back, and since focus went with the
 * lightbox, it moves to the thumbnail of the image that was on show.
 * @param {HTMLElement} root - The gallery's root element.
 * @param {number} index - The index of the image the lightbox showed last.
 */
export function closeLightbox(root, index) {
	if (!root) return;

	release(root);

	const active = document.activeElement;
	if (active && active !== document.body) return;

	root.querySelectorAll('.moka-media-gallery-item')[index]?.focus();
}

/**
 * Called when the gallery goes away with its lightbox still open, so it never renders closed.
 * Gives back the scroll lock and leaves focus alone, since the thumbnails are going too.
 * @param {HTMLElement} root - The gallery's root element.
 */
export function disposeLightbox(root) {
	if (root) release(root);
}
